sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCast.scala

```diff
@@ -0,0 +1,214 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.Literal.FalseLiteral
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+
+/**
+ * Unwrap casts in binary comparison operations with patterns like following:
+ *
+ * `BinaryComparison(Cast(fromExp, toType), Literal(value, toType))`
+ *   or
+ * `BinaryComparison(Literal(value, toType), Cast(fromExp, toType))`
+ *
+ * This rule optimizes expressions with the above pattern by either replacing the cast with simpler
+ * constructs, or moving the cast from the expression side to the literal side, which enables them
+ * to be optimized away later and pushed down to data sources.
+ *
+ * Currently this only handles cases where `fromType` (of `fromExp`) and `toType` are of integral
+ * types (i.e., byte, short, int and long). The rule checks to see if the literal `value` is
+ * within range `(min, max)`, where `min` and `max` are the minimum and maximum value of
+ * `fromType`, respectively. If this is true then it means we can safely cast `value` to `fromType`
+ * and thus able to move the cast to the literal side.
+ *
+ * If the `value` is not within range `(min, max)`, the rule breaks the scenario into different
+ * cases and try to replace each with simpler constructs.
+ *
+ * if `value > max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) === value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) <=> value` ==> false
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> if(isnull(exp), null, true)
+ *
+ * if `value == max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> exp == max
+ *  - `cast(exp, ty) === value` ==> exp == max
+ *  - `cast(exp, ty) <=> value` ==> exp == max
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> exp =!= max
+ *
+ * Similarly for the cases when `value == min` and `value < min`.
+ *
+ * Further, the above `if(isnull(exp), null, false)` is represented using conjunction
+ * `and(isnull(exp), null)`, to enable further optimization and filter pushdown to data sources.
+ * Similarly, `if(isnull(exp), null, true)` is represented with `or(isnotnull(exp), null)`.
+ */
+object UnwrapCast extends Rule[LogicalPlan] {
```

Comment:
Shall we rename to more precise one like `UnwrapIntegralCast` because this optimizer only handle `IntegralType`?

Reply:
You can keep the private function name `private def unwrapCast`.

Reply:
Thanks. If you are OK I'm slightly inclined to keep the `UnwrapCast` name though since this method is evolving and will expand to more types in follow-ups (should we name it to `UnwrapIntegralAndDateTypeCast` when we added `Date` later?). Plus the comments have already indicated that this only handle integral type at the moment.

Reply:
> Also, you may want to mention BinaryComparison like SimplifyBinaryComparison because this optimizer only handles BinaryComparison.

Hmm what do you mean? I did mention binary comparison in the comments, such as:
```
Unwrap casts in binary comparison operations with patterns like following
```

Reply:
The above comments are the same purpose. `UnwrapCast` is too broad and description is not enough always.
- `SimplifyBinaryComparison` has a clear meaning to the audience that it works only for `BinaryComparision`. Your optimize also does.

I don't think we can enumerate all types into the optimizer name like `UnwrapIntegralAndDateTypeCast`. We can choose more proper name at that context. The current `UnwrapCast` looks like a little over-claim to me.

Reply:
In the naming view point, I believe you had better add `Integral` or `BinaryComparision` to be clear.

Reply:
How about `UnwrapCastInBinaryComparison`?